scripts/pages/zoning/zoning.ts

```diff
@@ -0,0 +1,922 @@
+			this.initMenu();
+		}
+	}
+
+	static hideZoneMenu() {
+		if (this.panels.trackList?.Children().length) {
+			this.panels.trackList.RemoveAndDeleteChildren();
+		}
+	}
+
+	static toggleCollapse(container: Panel, expandIcon: Panel, collapseIcon: Panel) {
+		const shouldExpand = container.HasClass('hide');
+		container.SetHasClass('hide', !shouldExpand);
+		expandIcon.SetHasClass('hide', !shouldExpand);
+		collapseIcon.SetHasClass('hide', shouldExpand);
+		const parent = container.GetParent();
+		if (parent && parent.HasClass('zoning__tracklist-segment')) {
+			parent.SetHasClass('zoning__tracklist-segment--dark', shouldExpand);
+		}
+	}
+
+	static createTrackEntry(parent: Panel, entry: MainTrack | BonusTrack, name: string) {
+		const trackChildContainer = this.addTracklistEntry(parent, name, TracklistSnippet.TRACK, {
+			track: entry,
+			segment: null,
+			zone: null
+		});
+		if (trackChildContainer === null) return;
+		if (entry.zones.segments.length === 0) {
+			trackChildContainer.RemoveAndDeleteChildren();
+			(parent.FindChildTraverse('CollapseButton') as Panel).style.visibility = 'collapse';
+			return;
+		}
+
+		const trackSegmentContainer = trackChildContainer.FindChildTraverse('SegmentContainer') as Panel;
+		for (const [i, segment] of entry.zones.segments.entries()) {
+			const majorId = segment.name || `Segment ${i + 1}`;
+			const segmentChildContainer = this.addTracklistEntry(
+				trackSegmentContainer,
+				majorId,
+				TracklistSnippet.SEGMENT,
+				{
+					track: entry,
+					segment: segment,
+					zone: null
+				}
+			);
+			if (segmentChildContainer === null) continue;
+			if (segment.checkpoints.length === 0) {
+				(trackChildContainer.FindChildTraverse('CollapseButton') as Panel).style.visibility = 'collapse';
+				continue;
+			}
+
+			const segmentCheckpointContainer = segmentChildContainer.FindChildTraverse('CheckpointContainer') as Panel;
+			for (const [j, zone] of segment.checkpoints.entries()) {
+				const minorId = j ? `Checkpoint ${j}` : i ? 'Stage Start' : 'Start Zone';
+				this.addTracklistEntry(segmentCheckpointContainer, minorId, TracklistSnippet.CHECKPOINT, {
+					track: entry,
+					segment: segment,
+					zone: zone
+				});
+			}
+			if (!segment.cancel || segment.cancel.length === 0) continue;
+			for (const [j, zone] of segment.cancel.entries()) {
+				const cancelId = `Cancel ${j + 1}`;
+				this.addTracklistEntry(segmentCheckpointContainer, cancelId, TracklistSnippet.CHECKPOINT, {
+					track: entry,
+					segment: segment,
+					zone: zone
+				});
+			}
+		}
+
+		if (entry.zones.end) {
+			const trackEndZoneContainer = trackChildContainer.FindChildTraverse('EndZoneContainer') as Panel;
+			this.addTracklistEntry(trackEndZoneContainer, 'End Zone', TracklistSnippet.CHECKPOINT, {
+				track: entry,
+				segment: null,
+				zone: entry.zones.end
+			});
+		}
+	}
+
+	static addTracklistEntry(
+		parent: Panel,
+		name: string,
+		snippet: string,
+		object,
```

Comment:
Also I'd avoid using `object` is a name, whilst it's not a keyword in JS, it *is* a primitive type in TS, and `Object` is a JS builtin. If necessary use `obj`, but this would probably be better as `trackData` or something.